src/query.cc

```diff
@@ -0,0 +1,334 @@
+            Nan::Set(js_predicate, p_index++,
+                Nan::New<String>(
+                  ts_query_string_value_for_id(ts_query, predicate.value_id, &len)
+                ).ToLocalChecked());
+            break;
+          case TSQueryPredicateStepTypeDone:
+            Nan::Set(js_pattern_predicates, a_index++, js_predicate);
+            js_predicate = Nan::New<Array>();
+            p_index = 0;
+            break;
+        }
+      }
+    }
+
+    Nan::Set(js_predicates, pattern_index, js_pattern_predicates);
+  }
+
+  info.GetReturnValue().Set(js_predicates);
+}
+
+void Query::Matches(const Nan::FunctionCallbackInfo<Value> &info) {
+  Query *query = Query::UnwrapQuery(info.This());
+  const Tree *tree = Tree::UnwrapTree(info[0]);
+  uint32_t start_row    = Nan::To<uint32_t>(info[1]).ToChecked();
+  uint32_t start_column = Nan::To<uint32_t>(info[2]).ToChecked() << 1;
+  uint32_t end_row      = Nan::To<uint32_t>(info[3]).ToChecked();
+  uint32_t end_column   = Nan::To<uint32_t>(info[4]).ToChecked() << 1;
+
+  if (query == nullptr) {
+    Nan::ThrowError("Missing argument query");
+    return;
+  }
+
+  if (tree == nullptr) {
+    Nan::ThrowError("Missing argument tree");
+    return;
+  }
+
+  TSQuery *ts_query = query->query_;
+  TSNode rootNode = node_methods::UnmarshalNode(tree);
+  TSPoint start_point = {start_row, start_column};
+  TSPoint end_point = {end_row, end_column};
+  ts_query_cursor_set_point_range(ts_query_cursor, start_point, end_point);
+  ts_query_cursor_exec(ts_query_cursor, ts_query, rootNode);
+
+  Local<String> js_matches_string = Nan::New("matches").ToLocalChecked();
+  Local<String> js_nodes_string = Nan::New("nodes").ToLocalChecked();
+  Local<String> js_pattern_string = Nan::New("pattern").ToLocalChecked();
+  Local<String> js_captures_string = Nan::New("captures").ToLocalChecked();
+  Local<String> js_name_string = Nan::New("name").ToLocalChecked();
+  Local<String> js_node_string = Nan::New("node").ToLocalChecked();
+
+  Local<Array> js_matches = Nan::New<Array>();
+  unsigned match_index = 0;
+  vector<TSNode> nodes;
+  TSQueryMatch match;
+
+  while (ts_query_cursor_next_match(ts_query_cursor, &match)) {
+    Local<Array> js_captures = Nan::New<Array>();
+
+    for (uint16_t i = 0; i < match.capture_count; i++) {
+      const TSQueryCapture &capture = match.captures[i];
+
+      uint32_t capture_name_len = 0;
+      const char *capture_name = ts_query_capture_name_for_id(
+          ts_query, capture.index, &capture_name_len);
+
+      TSNode node = capture.node;
+      nodes.push_back(node);
+
+      Local<Value> js_capture = Nan::New(capture_name).ToLocalChecked();
+      Nan::Set(js_captures, i, js_capture);
+    }
+
+    Local<Object> js_match = Nan::New<Array>();
+    Nan::Set(js_match, 0,  Nan::New(match.pattern_index));
+    Nan::Set(js_match, 1, js_captures);
+    Nan::Set(js_matches, match_index++, js_match);
```

Comment:
If I'm understanding this right, the `js_matches` array will end up looking something like this, with a nested array for each match, and an array within that for the captures:

```js
[
  // first match
  [
    5,  // pattern index
    [   // 2 captures
      "foo",
      "bar"
    ]
  ],

  // second match
  [
    7,  // pattern index
    [   // 1 capture
      "baz"
    ]
  ],
]
```

What do you think about "flattening" the structure to something like this, in order to reduce allocations:

```js
[
  // first match
  5, // pattern index
  2, // number of captures,
  "foo",
  "bar,

  // second match
  7, // pattern index
  1, // number of captures
  "baz,
]
```

I think that it could potentially speed things up further.

If you're sick of messing with this and just want to be done with, I understand and I am cool with merging it as is, but if you are interested in optimizing further, I think it might be worthwhile.

Reply:
Sure, that's very interesting! I'll make the change.

Reply:
Besides once there I'll be just one pass away to doing it the WASM-bindings way, only difference will be C buffer vs JS array.